spec/praxis/request_stage_spec.rb

```diff
@@ -1,93 +1,181 @@
+
+        allow(controller_class).to receive(:before_callbacks).once.and_return(controller_before_callbacks)
+        allow(controller_class).to receive(:after_callbacks).once.and_return(controller_after_callbacks)
       end
-      it 'returns nil' do
-        expect( stage.execute ).to be(nil)
+
+      it "sets up and executes callbacks" do
+        expect(stage).to receive(:setup!)
+        expect(stage).to receive(:setup_deferred_callbacks!)
+        expect(stage).to receive(:execute)
+        expect(stage).to receive(:execute_callbacks).once.with(before_callbacks)
+        expect(stage).to receive(:execute_callbacks).once.with(after_callbacks)
+        expect(stage).to receive(:execute_controller_callbacks).once.with(controller_before_callbacks)
+        expect(stage).to receive(:execute_controller_callbacks).once.with(controller_after_callbacks)
+        #stage.run
       end
+
     end
-    context 'when one stage returns a Response' do
-      it "runs no further stages after that" do
-        double_stage_fail = double("fail stage")
-        expect(double_stage_ok).to receive(:run).once
-        expect(double_stage_fail).to receive(:run).once.and_return(Praxis::Responses::Ok.new)
-        stage.instance_variable_set(:@stages, [double_stage_ok, double_stage_fail, double_stage_ok])
-        stage.execute
+
+    context 'when the before execute_controller_callbacks return a Response' do
+      let(:action_name) { :show }
+      let(:response) { Praxis::Responses::Unauthorized.new }
+
+      before do
+        expect(stage).to receive(:execute_controller_callbacks).once.and_return(response)
       end
+
+
+      it 'does not call "execute"' do
+        expect(stage).to_not receive(:execute)
+      end
+
+      it 'does not execute any "after" callbacks' do
+        expect(stage).to_not receive(:after_callbacks)
+        expect(controller_class).to_not receive(:after_callbacks)
+      end
+
+      #it 'assigns controller.response' do
+      #
+      #  expect(controller).to receive(:response=).
+      #      with(response).twice.and_call_original
+      #
+      #end
+
     end
 
+    context 'with substages' do
+      before do
+        stage.stages.push(substage_1, substage_2, substage_3)
+      end
+
+
+      context 'when one returns a Response' do
+        let(:response) { Praxis::Responses::Unauthorized.new }
+
+        before do
+          expect(substage_1).to receive(:run).once
+          expect(substage_2).to receive(:run).once.and_return(response)
+        end
+
+        it 'runs no after callbacks' do
+          expect(stage).to_not receive(:after_callbacks)
+          expect(controller_class).to_not receive(:after_callbacks)
+        end
+
+        it 'assigns controller.response' do
+          # twice, because we do it once in #execute, and again in #run...
+          expect(controller).to receive(:response=).
+            with(response).twice.and_call_original
```

Comment:
I went back and forth about that, and went with this mostly because it plays nicer with all the other specs that use the `after` callback to run the code.